### PR TITLE
Added CommandPostprocessEvent

### PR DESCRIPTION
**Pull request by LoreSchaeffer**
Added CommandPostprocessEvent an event that is called after a command is executed with the result of the command

### CODE REVIEW

File: patches/api/0366-Added-CommandPostprocessEvent.patch [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+         }
+         // Paper end
+ 
++        boolean success = false; // Paper
++
+         try {
+             try (co.aikar.timings.Timing ignored = target.timings.startTiming()) { // Paper - use try with resources
+             // Note: we don't return the result of target.execute as thats success / failure, we return handled (true) or not handled (false)
+-            target.execute(sender, sentCommandLabel, Arrays.copyOfRange(args, 1, args.length));
++            success = target.execute(sender, sentCommandLabel, Arrays.copyOfRange(args, 1, args.length)); // Paper
```
Comment (ryantheleach) [inline patches/api/0366-Added-CommandPostprocessEvent.patch:109]: Read the comment directly above this line. this is NOT the success value! It's merely whether the command was handled or not handled.

**Comment by ryantheleach** [comment]
I think you need a slightly stricter contract for this to be useful to people.

You have provided neither the Success value as you call it in the API, nor the Result as you call it in the PR.

Can plugins rely on the output of this event never changing in the future? 

Like, is it a Informational only event, and thus is there a guarantee it won't be mutated in the future?

Or is the capability there for this to control CommandBlock decision making, by changing the success values?

I think you could add a lot more to this event, to make it more useful either way.



Your success value isn't whether the command succeeded, it's whether the command was handled.

`Success` has a definition in vanilla Minecraft, that I believe should be exposed here, if possible, in addition to whether the command was handled.

`Result` also has a definition in vanilla Minecraft, but this will be much harder to expose, as it can have varying types, including NBT. That said, it would provide a lot of compatibility if the result of any vanilla command was able to be represented.

**Comment by LoreSchaeffer** [comment]
You are right, i said that badly about the success / failure thing it was clearly handled / not handled.

My thought was to make an informational only event, to know when a command finish its execution and to know if it has been handled or not.

If you have any suggestions on how to make this better or more useful i can try to code them

**Comment by electronicboy** [comment]
post process is somewhat of a flawed name given that it's post-execution, not processing, processing was already done a good moment ago in the scheme of things

This event will not be fired if an exception is thrown by the command, which, creates many issues if this is solved in that you have no understanding of why it failed, etc

The nature of "Success" is also flawed here as already pointed out

**Comment by LoreSchaeffer** [comment]
> post process is somewhat of a flawed name given that it's post-execution, not processing, processing was already done a good moment ago in the scheme of things
> 
> This event will not be fired if an exception is thrown by the command, which, creates many issues if this is solved in that you have no understanding of why it failed, etc
> 
> The nature of "Success" is also flawed here as already pointed out

So changing the event to "CommandPostexecuteEvent" an handling the exception inside the dispatch method in SimpleCommandMap in this way could work better?

```java
boolean handled = false; // Paper

        try {
            try (co.aikar.timings.Timing ignored = target.timings.startTiming()) { // Paper - use try with resources
            // Note: we don't return the result of target.execute as thats success / failure, we return handled (true) or not handled (false)
            handled = target.execute(sender, sentCommandLabel, Arrays.copyOfRange(args, 1, args.length)); // Paper
            } // target.timings.stopTiming(); // Spigot // Paper
        } catch (CommandException ex) {
            server.getPluginManager().callEvent(new ServerExceptionEvent(new ServerCommandException(ex, target, sender, args))); // Paper
            // Paper start
            CommandPostexecuteEvent event = new CommandPostexecuteEvent(sender, target.getName(), Arrays.copyOfRange(args, 1, args.length), false, ex);
            server.getPluginManager().callEvent(event);
            // Paper end
            //target.timings.stopTiming(); // Spigot // Paper
            throw ex;
        } catch (Throwable ex) {
            //target.timings.stopTiming(); // Spigot // Paper
            String msg = "Unhandled exception executing '" + commandLine + "' in " + target;
            server.getPluginManager().callEvent(new ServerExceptionEvent(new ServerCommandException(ex, target, sender, args))); // Paper
            // Paper start
            CommandException ex1 = new CommandException(msg, ex);
            CommandPostexecuteEvent event = new CommandPostexecuteEvent(sender, target.getName(), Arrays.copyOfRange(args, 1, args.length), false, ex1);
            server.getPluginManager().callEvent(event);
            throw ex1;
            // Paper end
        }

        // Paper start
        CommandPostexecuteEvent event = new CommandPostexecuteEvent(sender, target.getName(), Arrays.copyOfRange(args, 1, args.length), handled);
        server.getPluginManager().callEvent(event);
        // Paper end
```

**Comment by Machine-Maker** [comment]
If something like this exists, it should be part of paper’s mojang api I think. And use brig’s ParseResults and what not.

**Comment by LoreSchaeffer** [comment]
> If something like this exists, it should be part of paper’s mojang api I think. And use brig’s ParseResults and what not.

I'm not very familiar with brigadier, any help is appreciated

**Pull request closed without merging**